### ISSUE TITLE
(client) ss="user-button sc-bdVaJa hJPaNh" data-r  (server) ss="user-button sc-bwzfXH eerwGk" data-r

### ISSUE THREAD

**Issue by andy1028** [styled-components/styled-components#1053]
describe：
server side render throw below error when use `extend`:

![image](https://user-images.githubusercontent.com/13925688/28821920-e6697302-76e9-11e7-86b9-8ebb71504ccf.png)

styleComponentCode:

```js
// button component

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

// 暴露行为 onClick
class Button extends React.Component {

  constructor(props) {
    super(props);
  }

  click() {
    this.props.onClick();
  }

  render() {
    const {
      className,
      onClick,
      children
    } = this.props;

    return (
      <div className={className} onClick={() => onClick()}>
        {children}
      </div>
    );
  }
}

```

```js
// button style component

export const DefaultButton = styled(Button)`
  font-size:14px;
  background-color: ${props => (props.disable ? '#d7d7d7' : '#f60')};
  padding: 8px 0;
  border: 0;
  width: ${props => (props.size ? Size[props.size] : '240px')};
  text-align: center;
  color: #fff;
  border-radius: 4px;
  &:hover{
    ${
      props => defauHoverFunc(props)
    }
  }
`;
```

```js
// use extend to extranl button

export const DottedButton = DefaultButton.extend`
  background-color: inherit;
  border: 1px solid ${props => (props.disable ? '#d7d7d7' : '#f60')};
  color: ${props => (props.disable ? '#d7d7d7' : '#f60')};
  &:hover{
    ${
      props => dotHoverFunc(props)
    }
  }
`;

```
run env:

"styled-components": "^2.1.0"
"next": "latest",
  "react": "15.6.1",
   "react-dom": "15.6.1",

**Comment by andy1028** [comment]
i try a sample demo
```js
const Button = styled.button`
  color: palevioletred;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid palevioletred;
  border-radius: 3px;
`;

const TomatoButton = Button.extend`
  color: tomato;
  border-color: tomato;
`;

const DemoPage = () => (
  <div>
    <Button>Normal Button</Button>
    <TomatoButton>Tomato Button</TomatoButton>
  </div>
);
export default DemoPage;
```
it has above error also

**Comment by andy1028** [comment]
i try update styled-components to 2.1.1 
it has above error also

**Comment by philpl** [comment]
With most setups for server-side rendering, the import order changes, which leads to different unique component ids on the client and server. To prevent this we have a babel plugin that assigns IDs that aren't based on the import order, but based on the filename and component name:

https://www.styled-components.com/docs/advanced#serverside-rendering

**Comment by philpl** [comment]
Btw, since I'm seeing this now. Generally speaking you'd want to avoid wrapping components with the `styled()` factory for your own components, and keep that for third-party ones or when it's absolutely necessary. This way you force yourself to keep your container/structural component logic separate and preserve nice features like our prop whitelist.

**Comment by andy1028** [comment]
thanks for reply，below is my .bablrc config

```json
{
  "presets": [
    "next/babel"
  ],
  "plugins": [
    ["styled-components", { "ssr": true, "displayName": false, "preprocess": true } ]
  ]
}
```
is has above error also

**Comment by wmertens** [comment]
@philpl I have the same problem with the same configuration. Even with the babel plugin, the names aren't consistent.

**Comment by wmertens** [comment]
For example, the initial HTML has this comment in the generated styles:

```css
/* sc-component-id: Page__PageWrapper-gxjOOq */
.Page__PageWrapper-gxjOOq {}
```

but React complains that
```
 (client) ="Page__PageWrapper-bnywRW caKErI" data-
 (server) ="Page__PageWrapper-gxjOOq kDpMb" data-r
```

How can I troubleshoot the rehydration?

**Comment by wmertens** [comment]
I thought I found the issue, there was one file that was creating different components with the same name, but even after fixing that the problem persists.

@philpl Can you reopen this one please?

**Comment by philpl** [comment]
@wmertens @andy1028 Sorry, I actually misread the description and overlooked the mention of `.extend`. This method breaks the deterministic ID generation right now, but it was fixed in this PR: https://github.com/styled-components/styled-components/pull/1044

It's not release yet though.

**Comment by mxstbr** [comment]
Ohh we should release that!

**Comment by philpl** [comment]
@mxstbr I wanted to wait for some of @schwers optimisations and release a minor, but we could release this as a patch already

**Comment by wmertens** [comment]
Hmm ok I think I'm hitting another case but let's see what that patch does…

**Comment by philpl** [comment]
@wmertens can you open a new issue with the code snippet that is causing this? Specifically, the resulting SSR critical CSS of this component, the **transpiled** components that match that `displayName` (possibly including the server and client code if you can), and some details on your hot reloading setup :)

**Comment by mxstbr** [comment]
Releases are cheap, let's push a patch out?

**Comment by philpl** [comment]
@mxstbr Ok, I'll release it and write up some quick release notes :+1:

**Comment by wmertens** [comment]
Ok, will create a new issue once I tested with the patch. I haven't been
able to locate a snippet so far, the failure location varies a little.

**Comment by braska** [comment]
I have same issue and I realized that error occurs in case of re-requiring compiled entrypoint (`src/serverRender.jsx`, for example).

In production environment `src/serverRender.jsx` is required once (on boot) in `index.js`. But in dev environment it's necessary re-require `src/serverRender.jsx` after each modification in sources.

I prepared demo application to reproduce issue: https://github.com/braska/styled-components-ssr-issue-demo

I have tried to add `console.log(identifiers);` to 73 line in `node_modules/styled-components/lib/models/StyledComponent.js` (https://github.com/styled-components/styled-components/blob/master/src/models/StyledComponent.js#L33) to determinate why server side environment generate wrong class names. I got unexpected results. See server console output:
```
--- Run app ---
webpack built 6b5796cb7bffd02ff59a in 3213ms

--- Open page in browser ---
{ app__Button: 1 }


--- Modify src/app.jsx ---
• Compiling...
webpack building...
✓ Compilation succeeded (850ms)

webpack built 8e0df5bae5184d86db97 in 850ms
{ app__Button: 2 }
```

Problem is that `app__Button` field contains `2` now. But in browser this field always contains `1`.
